cpp: thread: support for std::thread, std::mutex, std::condition_variable, etc

cfriedt opened the pull request:
This commit adds support for `std::thread` and `std::this_thread` when the configured C++ standard is >= C++11.

Fixes #25569

Accompanying `sdk-ng` PR: https://github.com/zephyrproject-rtos/sdk-ng/pull/735

```
SeaBIOS (version rel-1.15.0-0-g2dd4b9b3f840-prebuilt.qemu.org)
Booting from ROM..*** Booting Zephyr OS build zephyr-v3.0.0-1074-g90c018160af3  ***
version 201703
Running test suite std_thread_tests
===================================================================
START - test_this_thread_get_id
 PASS - test_this_thread_get_id in 0.2 seconds
===================================================================
START - test_this_thread_yield
 PASS - test_this_thread_yield in 0.1 seconds
===================================================================
START - test_this_thread_sleep_until
 PASS - test_this_thread_sleep_until in 0.1 seconds
===================================================================
START - test_this_thread_sleep_for
 PASS - test_this_thread_sleep_for in 0.1 seconds
===================================================================
START - test_thread_get_id
 PASS - test_thread_get_id in 0.1 seconds
===================================================================
START - test_thread_native_handle
 SKIP - test_thread_native_handle in 0.1 seconds
===================================================================
START - test_thread_hardware_concurrency
 PASS - test_thread_hardware_concurrency in 0.1 seconds
===================================================================
START - test_thread_join
 SKIP - test_thread_join in 0.1 seconds
===================================================================
START - test_thread_detach
 SKIP - test_thread_detach in 0.1 seconds
===================================================================
START - test_thread_swap
 SKIP - test_thread_swap in 0.1 seconds
===================================================================
Test suite std_thread_tests succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Notes:
- Tests pass for `CONFIG_ARCH_POSIX=y` and `CONFIG_ARCH_POSIX=n`
- Currently skipping a few tests when `CONFIG_ARCH_POSIX=n` as we cannot yet auto-allocate thread stacks
- Requires a couple of toolchain header modifications (described in comments / PR)
- Eventually would be ideal to bypass POSIX threads and use Zephyr native threads, but they are a convenient stepping stone

cfriedt commented:
The toolchain header `include/bits/gthr.h` was modified as follows:
```
#if _GLIBCXX___ANDROID__
#include <bits/gthr-posix.h>
#elif defined(__ZEPHYR__)
#include <bits/gthr-zephyr.h>
#else
#include <bits/gthr-default.h>
#endif
```

The file `include/bits/gthr-zephyr.h` is a copy of the upstream `gthr-posix.h` with the following additions. Normally, I would suggest keeping this in the toolchain, as this file is GPLv3. It is included here for reference purposes only.
```
#define __GTHREADS 1
#define __GTHREADS_CXX0X 1
#define _GTHREAD_USE_MUTEX_TIMEDLOCK 1
#define _GLIBCXX_USE_NANOSLEEP 1

struct __sFILE_fake {
	unsigned char *_p; /* current position in (some) buffer */
	int _r; /* read space left for getc() */
	int _w; /* write space left for putc() */
	short _flags; /* flags, below; this FILE is free if 0 */
	short _file; /* fileno, if Unix descriptor, else -1 */
	struct __sbuf _bf; /* the buffer (at least 1 byte, if !NULL) */
	int _lbfsize; /* 0 or -_bf._size, for inline putc */

	struct _reent *_data;
};

typedef long _off_t;
typedef unsigned long size_t;
typedef long ssize_t;
#define _fpos_t long
#define __FILE struct __sFILE_fake
typedef __builtin_va_list __gnuc_va_list;
#define PTHREAD_MUTEX_INITIALIZER {0}
#define PTHREAD_COND_INITIALIZER {0}

#include <posix/pthread.h>

#if ((defined(_LIBOBJC) || defined(_LIBOBJC_WEAK)) \
     || !defined(_GTHREAD_USE_MUTEX_TIMEDLOCK))
# include <posix/unistd.h>
```

cfriedt commented:
> @cfriedt is there any reason why this pull-req is not progressing forward?

The biggest reasons are
1. Zephyr has no way currently to dynamically allocate a thread stack (there is a draft pr for that)
2. The related toolchain integrations need to happen (should be doable)
3. Additional work may be necessary around `k_mutex` (iirc, there may be some static requirement for Zephyr mutex objects)

ShaharHD commented:
@cfriedt I want to add more relevant implementations to this PR like:
- `std::mutex`
- `std::timed_mutex`

Which I see were referecned in #45785 but are not addressed here.
But are all of the blocking items still relevant even if it's just the mutex?

cfriedt commented:
> @cfriedt I want to add more relevant implementations to this PR like:
> - `std::mutex`
> - `std::timed_mutex`
> 
> Which I see were referecned in #45785 but are not addressed here.

Well, we could, in theory, add C++ support now using the gthr_posix approach, but it would come with the caveat that C++ `std::thread` would be completely non-functional.

> But are all of the blocking items still relevant even if it's just the mutex?

From what I remember, no. That can be done.

Thoughts, @stephanosio ?

cfriedt commented:
* rebased (kind of insane to believe the last rev for this was 3 years ago 😅 )
* now you just need to change `bits/gthr.h` to include `bits/gthr-posix.h`

Currently fails, as it relies on automatic thread stack allocation via pthreads, but that is expected until #44379 and #44727 are merged.

cfriedt commented:
**TSC Meeting 2023-08-30:** Since this came up again, and since it was only relatively recently that this was unblocked (#44379, #44727), I had a few minutes so I thought I would give it another shot.

On my desk, using Zephyr SDK v0.16.1 and Zephyr commit f0c10d0332ec7cfc7582873c4cb4e3205efe2106.

With **unmodified** `gthr-posix.h` there are only a few minute tweaks needed to get this working.

The modifications I made were below.

1. include `<bits/gthr-posix.h>` instead of `<bits/gthr-default.h>` in `<bits/gthr.h>`
2. `#define _GLIBCXX_HAS_GTHREADS 1` in `bits/c++config.h`
3. `#define _GLIBCXX_USE_NANOSLEEP 1` in `bits/c++config.h`
4. Use `pthread_exit(nullptr)` (or `k_thread_abort(k_current_get())`) instead of `terminate()` in `std::terminate()`

For the last one, the (C) `terminate()` function is undocumented and seems like a gcc-ism. It's unfortunately, non-weak, so we can't easily override it.

The reason we cannot use`std::terminate()` is that the (C) `terminate()` ends up calling `abort()`.

That's OK on an OS with the concept of a process, because the OS would naturally keep going after a process terminates, but for a multi-threaded RTOS with a single-process view, it causes everything to stop.

I think at this point, it's probably safe to make a PR to `sdk-ng`.

Have not yet attempted any C++ threading with LLVM.

cc @nashif @carlescufi @yperess @stephanosio 

<img width="1190" alt="Screenshot 2023-08-30 at 5 38 34 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/3d666335-cc9d-49c9-b777-a2c96f6459da">

yperess commented:
@cfriedt I think the concern is that even if we get it working with the Zephyr toolchain that doesn't solve the issue for everyone right? Some projects use LLVM, others might have an internal GCC (can't use the Zephyr toolchain). While I think there's value in doing this for the Zephyr SDK. I do like the Pigweed solution better because it's a software solution that will work for everyone.

cfriedt commented:
> @cfriedt I think the concern is that even if we get it working with the Zephyr toolchain that doesn't solve the issue for everyone right?

Not sure I agree. Both GCC and LLVM have first-class support for POSIX in terms of C++ threads, so from that perspective, it's already like 95% of users.

It's relatively straightforward to configure GCC to do what we need. I just haven't gotten to LLVM yet - would be great to get some help there.

> Some projects use LLVM, others might have an internal GCC (can't use the Zephyr toolchain). 

Of course. The Zephyr SDK is just GCC at the end of the day, and we can provide our config as a reference for those who want to use custom GCC-based toolchains. Similarly, once we can support LLVM, then we'll provide the relevant configuration bits as well.

One step at a time though.

cfriedt commented:
The `cpp_synchronization` sample should probably be updated to use `std::binary_semaphore` or `std::counting_semaphore`. 

`std::mutex` and `std::condition_variable` are working fine (using gcc)

Some TODO items:
* modify Zephyr SDK config options to support C++ by default
* provide documentation for those configuring their own toolchains and / or 3rd-party toolchain vendors based on GCC
* get C++ working with LLVM

cfriedt commented:
Added tests for `std::mutex` and `std::condition_variable` as well.

<img width="656" alt="Screenshot 2023-10-08 at 1 29 41 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/824bc511-81ca-403c-a67c-c023077099ba">
<img width="616" alt="Screenshot 2023-10-08 at 1 28 24 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/fd01e5f9-7d79-48ce-bf39-5eae436f1883">
<img width="699" alt="Screenshot 2023-10-08 at 1 28 07 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/ce7507e6-d683-4616-8221-8ece657162d2">

valentinkorenblit commented:
Hi @cfriedt ,

I am trying to build a custom aarch64 toolchain with crosstool-ng and newlib and also c++ support to use with zephyr 3.4 and posix.

I've tried these steps so I can use std::mutex, std::thread, etc.

> 
> With **unmodified** `gthr-posix.h` there are only a few minute tweaks needed to get this working.
> 
> The modifications I made were below.
> 
>     1. include `<bits/gthr-posix.h>` instead of `<bits/gthr-default.h>` in `<bits/gthr.h>`
> 
>     2. `#define _GLIBCXX_HAS_GTHREADS 1` in `bits/c++config.h`
> 
>     3. `#define _GLIBCXX_USE_NANOSLEEP 1` in `bits/c++config.h`
> 
>     4. Use `pthread_exit(nullptr)` (or `k_thread_abort(k_current_get())`) instead of `terminate()` in `std::terminate()`
>
 
Do you confirm that the following steps are done after toolchain's install step? because if doing this modifications before building gcc the build breaks because of undefined pthread objects, such as:


 `src/gcc/libgcc/gthr-posix.h:102:9: error: 'pthread_once' undeclared here (not in a function); did you mean 'pthread_once_t'? `
 `src/gcc/libgcc/gthr-posix.h:103:9: error: 'pthread_getspecific' undeclared here (not in a function) `
 `src/gcc/libgcc/gthr-posix.h:106:9: error: 'pthread_create' undeclared here (not in a function) `

However, if I patch c++config.h and gthr.h after the install step, I manage to build the toolchain and when building some zephyr code that uses std::mutex/condition_variable/thread it compiles fine but of course fails at link time because the cc files are all
flagged with  `#ifdef _GLIBCXX_HAS_GTHREADS ` (which is not defined when building libstdc++-v3)

Do you spot if I am missing something here?

Thanks in advance,

Valentin

cfriedt commented:
Hi @valentinkorenblit 

> Do you confirm that the following steps are done after toolchain's install step? because if doing this modifications before building gcc the build breaks because of undefined pthread objects, such as:

That was the way that I tested my code, yes.

However, the proper way to configure a custom GCC + Newlib toolchain would be to apply any necessary patches and then use the appropriate arguments to `./configure`.

There are great projects out there to make it significantly easier - yocto, buildroot, gentoo, crossdev-ng, of course.

> However, if I patch c++config.h and gthr.h after the install step, I manage to build the toolchain and when building some zephyr code that uses std::mutex/condition_variable/thread it compiles fine but of course fails at link time because the cc files are all
> flagged with  `#ifdef _GLIBCXX_HAS_GTHREADS ` (which is not defined when building libstdc++-v3)

Exactly - so you would need to apply patches first, and then use the appropriate arguments to `./configure`.

> Do you spot if I am missing something here?

Nope - you're basically a couple of steps ahead of me due to lack of bandwidth on my part.

If you happen to find the correct arguments to `./configure`, can you please note them down for us?

This would be really helpful to add to a page under "Beyond the Getting Started Guide", and similar steps would be great to document for configuring LLVM.

Eventually, it would be great to stabilize this and just upstream the configuration to both GCC and LLVM, but that is a bit further down the road.

> Thanks in advance

Likewise 😀

cfriedt commented:
Reproducing again, the patch below is all that is required to modify an _existing_ Zephyr SDK toolchain install (to use C++).

I mitigated item 4 by calling `std::set_terminate()` from a `SYS_INIT()` macro (so no longer need to patch `c++config.h` with `pthread_exit()`.

```diff
diff -Nuar riscv64-zephyr-elf_unomodified/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/c++config.h riscv64-zephyr-elf/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/c++config.h
--- riscv64-zephyr-elf_unomodified/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/c++config.h	2023-11-15 06:45:56.000000000 -0500
+++ riscv64-zephyr-elf/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/c++config.h	2023-12-01 14:54:24.261480902 -0500
@@ -1688,7 +1688,7 @@
 #define _GLIBCXX_FULLY_DYNAMIC_STRING 0
 
 /* Define if gthreads library is available. */
-/* #undef _GLIBCXX_HAS_GTHREADS */
+#define _GLIBCXX_HAS_GTHREADS 1
 
 /* Define to 1 if a full hosted library is built, or 0 if freestanding. */
 #define _GLIBCXX_HOSTED 1
@@ -1820,7 +1820,7 @@
 /* #undef _GLIBCXX_USE_LSTAT */
 
 /* Defined if nanosleep is available. */
-/* #undef _GLIBCXX_USE_NANOSLEEP */
+#define _GLIBCXX_USE_NANOSLEEP 1
 
 /* Define if NLS translations are to be used. */
 /* #undef _GLIBCXX_USE_NLS */
diff -Nuar riscv64-zephyr-elf_unomodified/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/gthr.h riscv64-zephyr-elf/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/gthr.h
--- riscv64-zephyr-elf_unomodified/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/gthr.h	2023-11-15 06:45:56.000000000 -0500
+++ riscv64-zephyr-elf/riscv64-zephyr-elf/include/c++/12.2.0/riscv64-zephyr-elf/rv64imac_zicsr_zifencei/lp64/medany/bits/gthr.h	2023-12-01 13:24:00.184251743 -0500
@@ -145,7 +145,7 @@
 #define _GLIBCXX_GTHREAD_USE_WEAK 1
 #endif
 #endif
-#include <bits/gthr-default.h>
+#include <bits/gthr-posix.h>
 
 #ifndef _GLIBCXX_HIDE_EXPORTS
 #pragma GCC visibility pop
```

cfriedt commented:
Links to relevant SDK-related PRs
-[ ] zephyrproject-rtos/gcc#25
-[ ] zephyrproject-rtos/sdk-ng#722

cfriedt commented:
<img width="882" alt="Screenshot 2023-12-01 at 4 44 38 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/dcdbe176-44c1-4219-afa2-dffa68a9f079">

cfriedt commented:
Slightly different direction at this point - namely using the ISO C11 threads API to support ISO C++11 threads in gcc. Just performing some final testing.

https://github.com/zephyrproject-rtos/sdk-ng/pull/735

cfriedt commented:
> I am trying to build a custom aarch64 toolchain with crosstool-ng and newlib and also c++ support to use with zephyr 3.4 and posix.

@valentinkorenblit - check out the changes in this PR as well as the submodule changes (it renders much better on the desktop at the moment).

https://github.com/zephyrproject-rtos/sdk-ng/pull/735

That should allow you quite easily to build a cross toolchain with C++ support using either newlib or picolibc.

valentinkorenblit commented:
> > I am trying to build a custom aarch64 toolchain with crosstool-ng and newlib and also c++ support to use with zephyr 3.4 and posix.
> 
> @valentinkorenblit - check out the changes in this PR as well as the submodule changes (it renders much better on the desktop at the moment).
> 
> [zephyrproject-rtos/sdk-ng#735](https://github.com/zephyrproject-rtos/sdk-ng/pull/735)
> 
> That should allow you quite easily to build a cross toolchain with C++ support using either newlib or picolibc.

Hi @cfriedt , unfortunately I won't be able to test it because my target does not have an FPU and I had to do a lot more tricks to get it working. I finished this a few months ago but it is not something worth upstreaming.

cfriedt committed:
tests: cpp: tests for std::thread and std::this_thread

This commit adds tests for std::thread and std::this_thread.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: lib: cpp: add tests for std::condition_variable

Add tests for the ISO C++11 std::condition_variable.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
cpp: thread: support for std::thread and std::this_thread

This commit adds support for std::thread and std::this_thread
when the configured C++ standard is >= C++11.

The implementation uses POSIX threads under the hood.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: lib: cpp: add tests for std::mutex and friends

Add tests for ISO C++ std::mutex.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
<img width="734" alt="Screenshot 2024-03-09 at 1 19 55 PM" src="https://github.com/zephyrproject-rtos/sdk-ng/assets/983494/2870ed53-91f3-4136-a4bb-b4e2055576b2">
<img width="628" alt="Screenshot 2024-03-09 at 1 20 31 PM" src="https://github.com/zephyrproject-rtos/sdk-ng/assets/983494/533de7a4-4d6e-415d-a409-a7a4d004e293">
<img width="672" alt="Screenshot 2024-03-09 at 1 20 55 PM" src="https://github.com/zephyrproject-rtos/sdk-ng/assets/983494/8a8edba9-163a-4bfb-9b0d-8a3af1317ccf">
<img width="529" alt="Screenshot 2024-03-09 at 1 21 23 PM" src="https://github.com/zephyrproject-rtos/sdk-ng/assets/983494/7f50146e-5843-4ab3-8535-668f9e3cfe9b">

cfriedt commented:
marked `DNM` until `sdk-ng` changes are accepted.

cfriedt commented:
cc @XenuIsWatching

cfriedt commented:
Currently, eval SDK Toolchain components are ready to download for evaluation purposes 

* aarch64-zephyr-elf
* arm-zephyr-eabi
* riscv64-zephyr-elf
* x86_64-zephyr-elf

The x86_64 and riscv64 toolchain components also support the 32-bit variants for those ISAs

cfriedt commented:
This is just waiting on a candidate toolchain or CI image to become available.

v0.16.6 does have the necessary toolchain changes (aside from having to enable c11 threads) so as long as we can point Zephyr CI at a different image with that toolchain it can run through automated testing.

@stephanosio - is there a way that this can be pulled into the CI run you had going

AfflatusX commented:
Hi @cfriedt - I patched the 16.6 build to give the threading support a try. I was able to build/link my firmware successfully with the patch, thanks to your work!

I noticed that whenever __cxxabiv1::__cxa_guard_acquire() is invoked, it would throw exception though. I'm building against QEMU MPS3_AN547 target (cortex M55). When running LLDB against the elf I saw following callstack:

`
(lldb) bt
* thread #1, stop reason = breakpoint 1.1
  * frame #0: 0x10008160 zephyr.elf`__cxa_throw
    frame #1: 0x1000446c zephyr.elf`(anonymous namespace)::mutex_wrapper::mutex_wrapper() (.isra.0) + 44
    frame #2: 0x100044a0 zephyr.elf`__cxa_guard_acquire + 24
    frame #3: 0x100015ce zephyr.elf`main at main.cpp:27:33
    frame #4: 0x100031f0 zephyr.elf`bg_thread_main(unused1=0x00000000, unused2=0x00000000, unused3=0x00000000) at init.c:459:8
    frame #5: 0x10001672 zephyr.elf`z_thread_entry(entry=(zephyr.elf`bg_thread_main + 1 at init.c:424:20), p1=0x00000000, p2=0x00000000, p3=0x00000000) at thread_entry.c:48:2
    frame #6: 0x10002612 zephyr.elf`arch_switch_to_main_thread(main_thread=<unavailable>, stack_ptr=0x00000000, _main=<unavailable>) at thread.c:575:2
`

I want to debug this further, but as you can see the symbols for anything in stdlibc++ is missing. Two questions for you:

1. have you run into this or any guess on why this might be happening?
2. any suggestion on how can I load the libstdc++'s symbol into lldb so I can debug further?

Again, thanks for the great work here!

cfriedt commented:
Hi @AfflatusX - 

There is a known issue and fix, documented here

https://github.com/zephyrproject-rtos/sdk-ng/issues/751
https://github.com/zephyrproject-rtos/gcc/pull/31

I'm not sure what release this will be integrated into in the Zephyr SDK.

@stephanosio is the C++ maintainer and SDK maintainer, and might have a better idea.

AfflatusX commented:
@cfriedt - Thanks for the quick reply! I can fork/patch your PR in the meantime. However, when I pull the v0.16.6 branch and try to build the toolchain using `contrib/linux_build_toochain.sh` it failed at the GCC step with following output:

```
[ALL  ]    Using host-linux.o for host machine hooks.
[ALL  ]    c11 is an unknown thread package
[ERROR]    make[2]: *** [Makefile:4565: configure-gcc] Error 1
[ALL  ]    make[2]: Leaving directory '/home/txie/zephyr-sdk-build/build-arm-zephyr-eabi/.build/arm-zephyr-eabi/build/build-cc-gcc-final'
[ERROR]    make[1]: *** [Makefile:1038: all] Error 2
[ALL  ]    make[1]: Leaving directory '/home/txie/zephyr-sdk-build/build-arm-zephyr-eabi/.build/arm-zephyr-eabi/build/build-cc-gcc-final'
```

I tried to search for the message `c11 is an unknown thread package` but couldn't find anything useful. Any thoughts?

cfriedt commented:
> @cfriedt - Thanks for the quick reply! I can fork/patch your PR in the meantime. However, when I pull the v0.16.6 branch and try to build the toolchain using `contrib/linux_build_toochain.sh` it failed at the GCC step with following output:
> 
> I tried to search for the message `c11 is an unknown thread package` but couldn't find anything useful. Any thoughts?

I would guess that some changes were reverted in `sdk-ng` and possibly elsewhere.

AfflatusX commented:
> > @cfriedt - Thanks for the quick reply! I can fork/patch your PR in the meantime. However, when I pull the v0.16.6 branch and try to build the toolchain using `contrib/linux_build_toochain.sh` it failed at the GCC step with following output:
> > I tried to search for the message `c11 is an unknown thread package` but couldn't find anything useful. Any thoughts?
> 
> I would guess that some changes were reverted in `sdk-ng` and possibly elsewhere.

hmm, I'm sitting on the v0.16.6 tag:
```
txie@HABOR:~/sdk-ng-atomic$ git log
commit 2d488e6bef4bc258c0061f12d33ac083acd9ae1d (HEAD -> threads, tag: v0.16.6)
Author: Stephanos Ioannidis <root@stephanos.io>
Date:   Mon Apr 29 14:59:17 2024 +0900

    VERSION: Bump to 0.16.6
``` 

looking at the CI history, this commit passed the full CI suite, so it should build successfully right?

Is there a branch contains your latest patches with thread enabled I can give it a try locally?

cfriedt commented:
This needs a rebase and re-rev of SDK changes to be made.
